File: code_review_mcp/src/config/loader.ts
/**
 * Loads and validates the YAML configuration file.
 */

import { readFileSync, existsSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { parse as parseYaml } from "yaml";
import type { AppConfig, ProjectConfig } from "./types.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

/** Default config path: <project_root>/config.yaml */
const DEFAULT_CONFIG_PATH = resolve(__dirname, "..", "..", "config.yaml");

/**
 * Load the config from disk. Falls back to the default path if none is provided.
 */
export function loadConfig(configPath?: string): AppConfig {
  const cfgPath = configPath ?? DEFAULT_CONFIG_PATH;

  if (!existsSync(cfgPath)) {
    throw new Error(
      `Config file not found at: ${cfgPath}\n` +
        `Copy config.example.yaml to config.yaml and fill in your credentials.`
    );
  }

  const raw = readFileSync(cfgPath, "utf-8");
  const parsed = parseYaml(raw) as AppConfig;

  validateConfig(parsed);

  return parsed;
}

function validateConfig(cfg: AppConfig): void {
  const bb = cfg.auth?.bitbucket;
  if (!bb) {
    throw new Error("Config validation error: auth.bitbucket is required.");
  }

  const hasToken = Boolean(bb.email?.trim()) && Boolean(bb.api_token?.trim());
  const hasAppPassword = Boolean(bb.username?.trim()) && Boolean(bb.app_password?.trim());

  if (!hasToken && !hasAppPassword) {
    throw new Error(
      "Config validation error: provide either:\n" +
        "  - auth.bitbucket.email + auth.bitbucket.api_token (Atlassian API Token, recommended)\n" +
        "  - auth.bitbucket.username + auth.bitbucket.app_password (legacy App Password)"
    );
  }

  if (!Array.isArray(cfg.projects) || cfg.projects.length === 0) {
    throw new Error(
      "Config validation error: at least one entry in 'projects' is required."
    );
  }

  for (const p of cfg.projects) {
    if (!p.repo_slug) {
      throw new Error(
        "Config validation error: every project entry must have a 'repo_slug'."
      );
    }
  }
}

/**
 * Find the project config that matches a given workspace/repo-slug.
 * Falls back to a wildcard "*" entry if no exact match.
 */
export function findProjectConfig(
  config: AppConfig,
  workspace: string,
  repoSlug: string
): ProjectConfig | null {
  const fullSlug = `${workspace}/${repoSlug}`;

  // Exact match
  const exact = config.projects.find(
    (p) => p.repo_slug.toLowerCase() === fullSlug.toLowerCase()
  );
  if (exact) return exact;

  // Wildcard fallback
  const wildcard = config.projects.find((p) => p.repo_slug === "*");
  return wildcard ?? null;
}

/**
 * Read a file from disk (used for loading instruction / guideline markdown files).
 * Returns null if the path is null/undefined or file doesn't exist.
 */
export function loadFileContent(filePath: string | null | undefined): string | null {
  if (!filePath) return null;

  const resolved = resolve(filePath);
  if (!existsSync(resolved)) {
    console.error(`Warning: instruction file not found: ${resolved}`);
    return null;
  }

  return readFileSync(resolved, "utf-8");
}

File: integration_deployment_tools_mcp/internal/config/loader.go
package config

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"gopkg.in/yaml.v3"
)

// Load reads and validates the YAML configuration file.
func Load(configPath string) (*AppConfig, error) {
	if configPath == "" {
		exe, err := os.Executable()
		if err != nil {
			return nil, fmt.Errorf("resolve executable path: %w", err)
		}
		configPath = filepath.Join(filepath.Dir(exe), "config.yaml")
	}

	data, err := os.ReadFile(configPath)
	if err != nil {
		return nil, fmt.Errorf(
			"config file not found at %s\nCopy config.example.yaml to config.yaml and fill in your credentials",
			configPath,
		)
	}

	var cfg AppConfig
	if err := yaml.Unmarshal(data, &cfg); err != nil {
		return nil, fmt.Errorf("parse config: %w", err)
	}

	if err := validate(&cfg); err != nil {
		return nil, err
	}

	return &cfg, nil
}

func validate(cfg *AppConfig) error {
	bb := cfg.Auth.Bitbucket
	hasToken := strings.TrimSpace(bb.Email) != "" && strings.TrimSpace(bb.APIToken) != ""
	hasAppPassword := strings.TrimSpace(bb.Username) != "" && strings.TrimSpace(bb.AppPassword) != ""

	if !hasToken && !hasAppPassword {
		return fmt.Errorf(
			"config validation error: provide either:\n" +
				"  - auth.bitbucket.email + auth.bitbucket.api_token (Atlassian API Token, recommended)\n" +
				"  - auth.bitbucket.username + auth.bitbucket.app_password (legacy App Password)",
		)
	}

	if len(cfg.Repositories) == 0 {
		return fmt.Errorf("config validation error: at least one entry in 'repositories' is required")
	}

	for i, r := range cfg.Repositories {
		if strings.TrimSpace(r.RepoSlug) == "" {
			return fmt.Errorf("config validation error: repositories[%d].repo_slug is required", i)
		}
	}

	return nil
}

// FindRepoConfig looks up a repository config by its full slug (workspace/repo).
// Falls back to a wildcard ("*") entry if no exact match is found.
func FindRepoConfig(cfg *AppConfig, slug string) *RepoConfig {
	lower := strings.ToLower(slug)
	for i := range cfg.Repositories {
		if strings.ToLower(cfg.Repositories[i].RepoSlug) == lower {
			return &cfg.Repositories[i]
		}
	}
	for i := range cfg.Repositories {
		if cfg.Repositories[i].RepoSlug == "*" {
			return &cfg.Repositories[i]
		}
	}
	return nil
}

// FindRepoConfigByParts looks up a repository config by workspace and repo separately.
func FindRepoConfigByParts(cfg *AppConfig, workspace, repo string) *RepoConfig {
	return FindRepoConfig(cfg, workspace+"/"+repo)
}
